src/main/scala/mesosphere/marathon/core/health/impl/AppHealthCheckActor.scala

```diff
@@ -0,0 +1,157 @@
+    /**
+      * Map of health check definitions of all applications
+      */
+    var healthChecks: mutable.Map[ApplicationKey, Set[HealthCheck]] = mutable.Map.empty
+
+    /**
+      *  Map of results of all health checks for all applications.
+      *  Results are optional, therefore the global health status of an instance
+      *  is either:
+      *    unknown (if some results are still missing)
+      *    healthy (if all results are known and healthy)
+      *    not healthy (if all results are known and at least one is unhealthy)
+      */
+    var healthCheckStates: mutable.Map[InstanceKey, Map[HealthCheck, Option[Health]]] =
+      mutable.Map.empty
+
+    private def computeGlobalHealth(instanceHealthResults: Map[HealthCheck, Option[Health]]): Option[Boolean] = {
+      val isHealthAlive = (health: Option[Health]) => health.fold(false)(_.alive)
+      val isHealthUnknown = (health: Option[Health]) => health.isEmpty
+
+      if (instanceHealthResults.values.forall(isHealthAlive))
+        Some(true)
+      else if (instanceHealthResults.values.exists(isHealthUnknown))
+        Option.empty[Boolean]
+      else
+        Some(false)
+    }
+
+    def addHealthCheck(applicationKey: ApplicationKey, healthCheck: HealthCheck): Unit = {
+      healthChecks.update(applicationKey, healthChecks.getOrElse(applicationKey, Set.empty) + healthCheck)
+      logger.debug(s"Add health check $healthCheck to instance appId:${applicationKey.appId} version:${applicationKey.version}")
+    }
+
+    def removeHealthCheck(applicationKey: ApplicationKey, healthCheck: HealthCheck): Unit = {
+      logger.debug(s"Remove health check $healthCheck from instance appId:${applicationKey.appId} version:${applicationKey.version}")
+
+      val healthChecksAfterRemoval = healthChecks.getOrElse(applicationKey, Set.empty) - healthCheck
+      if (healthChecksAfterRemoval.isEmpty)
+        healthChecks.remove(applicationKey)
+      else
+        healthChecks.update(applicationKey, healthChecksAfterRemoval)
+
+      healthCheckStates = healthCheckStates.map(kv => {
+        val newHealthChecks = kv._2.filter(x => x._1 != healthCheck)
+        (kv._1, newHealthChecks)
+      }).filter(kv => kv._2.nonEmpty)
+    }
+
+    def purgeHealthCheckStatuses(toPurge: Seq[(InstanceKey, HealthCheck)]): Unit = {
+      toPurge.foreach(hc => {
+        val instanceKey = hc._1
+        val healthCheck = hc._2
+        val newInstanceHealthChecks = healthCheckStates.getOrElse(instanceKey, Map.empty) - healthCheck
+        if (newInstanceHealthChecks.isEmpty)
+          healthCheckStates.remove(instanceKey)
+        else
+          healthCheckStates.update(instanceKey, newInstanceHealthChecks)
+      })
+    }
+
+    def updateHealthCheckStatus(appKey: ApplicationKey, healthCheck: HealthCheck, health: Health,
+      notifier: (Option[Boolean] => Unit)): Unit = {
+      val hcDefinitions = healthChecks.getOrElse(appKey, Set.empty)
+      if (hcDefinitions.contains(healthCheck)) {
```

Comment:
Perhaps this reads little better?

```
healthChecks.get(appKey) match {
  case Some(hcDefinitions) if hcDefinitions.contains(healthCheck) =>
    ...
  case _ =>
    logger.warn(s"Status of $healthCheck health check changed but it does not exist in inventory")
}
```

(It is honestly fine is written, though :-))